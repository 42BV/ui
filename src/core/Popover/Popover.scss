.tippy-popover.tippy-box {
  color: $body-color;
  background-color: $popover-bg;

  .tippy-content {
    padding: 0;
  }

  .tippy-arrow {
    color: $popover-border-color !important;
  }
}
