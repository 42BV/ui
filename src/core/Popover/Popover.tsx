import React, { CSSProperties, useEffect, useState } from 'react';
import Tippy from '@tippyjs/react';
import { TippyPlacement } from '../types';

type Props = {
  /**
   * Optionally whether the popover is currently open,
   * when undefined, which is the default, the popover will show
   * when hovered over the target(children). When providing a boolean
   * you can take over when the popover is shown from outside the
   * Popover component.
   */
  isOpen?: boolean;

  /**
   * Optionally whether the popover should open on click instead of hover.
   */
  openOnClick?: boolean;

  /**
   * Optional callback that gets triggered when the tag is clicked.
   * When openOnClick is true, this component will take care of its visibility
   * state, unless onClick is defined.
   * This should be used when wanting to take complete control over the popover.
   */
  onClick?: () => void;

  /**
   * Optionally callback that gets triggered when clicked outside the popover.
   * Is useful for when wanting to take complete control over the popover.
   */
  onClickOutside?: () => void;

  /**
   * Content shown inside the popover.
   */
  children: React.ReactNode;

  /**
   * Target component that, when hovered, will trigger the popover to show up.
   * The target(children) of the popover are wrapped into a div.
   * This is a bypass to not have to forward the ref to the DOM node.
   */
  target: React.ReactNode;

  /**
   * Optional alignment relative to the target where the popover will be shown.
   */
  placement?: TippyPlacement;

  /**
   * Optional offset that the popover will show up relative from the target.
   */
  offset?: number;

  /**
   * Optional distance that the popover will show up relative from the target.
   */
  distance?: number;

  /**
   * Optionally the tag wrapping the children.
   * Default value is a span.
   */
  tag?: 'span' | 'div' | 'button';

  /**
   * Optional className that is added to the Wrapper component
   * Allowing you to add classes like margins and padding that would otherwise get lost
   * by the wrapping of the children inside the CustomTag.
   */
  className?: string;

  /**
   * Optional CSS properties that are added to the Wrapper component
   * Allowing you to add CSS properties that would otherwise get lost
   * by the wrapping of the children inside the CustomTag.
   */
  style?: CSSProperties;

  /**
   * Optionally set the maximum width of the popover.
   */
  maxWidth?: number;
};

/**
 * Bootstrap-like Popover component based on the Tippy.js library.
 */
export function Popover({
  children,
  placement = 'top',
  target,
  offset = 0,
  distance = 7,
  tag = 'span',
  className,
  isOpen,
  openOnClick,
  onClick,
  onClickOutside,
  style,
  maxWidth
}: Props) {
  const Tag = tag;

  const [visible, setVisible] = useState(openOnClick ? !!isOpen : undefined);

  useEffect(() => {
    setVisible(openOnClick ? !!isOpen : undefined);
  }, [openOnClick, isOpen]);

  function tagClicked() {
    if (onClick) {
      onClick();
    } else if (openOnClick) {
      setVisible(!visible);
    }
  }

  return (
    <Tippy
      visible={visible}
      onClickOutside={onClickOutside}
      className="border-0 tippy-popover"
      content={children}
      placement={placement}
      offset={[offset, distance]}
      interactive={true}
      zIndex={1049} // One level below bootstrap's modal
      maxWidth={maxWidth}
    >
      <Tag
        className={className}
        style={style}
        tabIndex={0}
        role="button"
        onClick={tagClicked}
      >
        {target}
      </Tag>
    </Tippy>
  );
}
