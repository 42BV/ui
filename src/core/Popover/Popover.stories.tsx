import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import {
  Alert,
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  Col,
  ListGroup,
  ListGroupItem,
  Row
} from 'reactstrap';

import { Popover } from './Popover';
import { Tag } from '../Tag/Tag';

import { ConfirmButton } from '../ConfirmButton/ConfirmButton';
import { Button } from '../Button/Button';
import { OpenCloseModal } from '../OpenCloseModal/OpenCloseModal';

storiesOf('core/Popover', module)
  .addParameters({ component: Popover })
  .addDecorator((Story) => (
    <>
      <Alert color="warning" className="mb-4">
        <p className="mb-0">To be able to use Popover, you have to add @tippyjs/react and tippy.js to your dependencies:</p>
        <code>npm install --save @tippyjs/react tippy.js</code>
        <p className="mb-0 mt-2">You also have to add the stylesheet to your project</p>
        <code>@import &apos;~tippy.js/dist/tippy.css&apos;;</code>
      </Alert>
      <Story />
    </>
  ))
  .add('default', () => (
    <div className="d-flex flex-column">
      <Row className="my-3">
        <Col className="d-flex justify-content-around align-items-center">
          <Popover target={<Tag color="danger" text="Hover me!" />}>
            <TinyCrud />
          </Popover>

          <Popover target={<h5>Can be wrapped around any component</h5>}>
            <TinyCrud />
          </Popover>

          <Popover target="Plain text">
            <NiceCard />
          </Popover>

          <Popover target={<Button>Hover this button!</Button>}>
            <TinyCrud />
          </Popover>
        </Col>
      </Row>
    </div>
  ))

  .add('taking control', () => {
    const [isOpen, setIsOpen] = useState(false);

    return (
      <div className="d-flex flex-column">
        <Row className="my-3">
          <Col className="d-flex justify-content-around align-items-center">
            Status: {isOpen ? 'opened' : 'closed'}
            <Popover isOpen={isOpen} target="Open">
              <TinyCrud />
            </Popover>
            <Button onClick={() => setIsOpen(!isOpen)}>Show / hide</Button>
          </Col>
          <Col>
            <p>
              Note: you can take complete controll over the Popover by using the{' '}
              <pre className="d-inline text-info">isOpen</pre> prop. Once you
              make it <pre className="d-inline text-info">true</pre> or{' '}
              <pre className="d-inline text-info">false</pre> the hover behavior
              will be disabled.
            </p>
          </Col>
        </Row>
      </div>
    );
  })

  .add('on click outside', () => {
    const [isOpen, setIsOpen] = useState(false);

    return (
      <div className="d-flex flex-column">
        <Row className="my-3">
          <Col className="d-flex justify-content-around align-items-center">
            Status: {isOpen ? 'opened' : 'closed'}
            <Popover
              isOpen={isOpen}
              onClickOutside={() => setIsOpen(false)}
              target="Open"
            >
              <NiceCard />
            </Popover>
            <Button onClick={() => setIsOpen(true)}>Show</Button>
          </Col>
          <Col>
            <p>
              Note: you can take complete controll over the Popover by using the{' '}
              <pre className="d-inline text-info">isOpen</pre> prop. Once you
              make it <pre className="d-inline text-info">true</pre> or{' '}
              <pre className="d-inline text-info">false</pre> the hover behavior
              will be disabled.
            </p>

            <p>
              In combination with{' '}
              <pre className="d-inline text-info">onClickOutside</pre> you can
              close the popover when clicked anywhere outside the popover.
            </p>
          </Col>
        </Row>
      </div>
    );
  })

  .add('alignment', () => (
    <>
      <h6>Alignment</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Popover
            target={<Tag color="danger" text="Hover me!" />}
            placement="bottom"
          >
            Placement bottom
          </Popover>

          <Popover
            target={<Tag color="warning" text="Hover me!" />}
            placement="left"
          >
            Placement left
          </Popover>
          <Popover
            target={<Tag color="primary" text="Hover me!" />}
            placement="right"
          >
            Placement right
          </Popover>

          <Popover
            target={<Tag color="success" text="Hover me!" />}
            placement="top"
          >
            Placement top
          </Popover>
        </Col>
      </Row>
      <hr />

      <h6>Alignment-modifier</h6>
      <Row className="mt-3">
        <Col className="d-flex justify-content-around">
          <Popover
            target={<Button> Hover me! </Button>}
            placement="right-start"
          >
            right-start
          </Popover>

          <Popover target={<Button> Hover me! </Button>} placement="right">
            right
          </Popover>

          <Popover target={<Button> Hover me! </Button>} placement="right-end">
            right-end
          </Popover>
        </Col>
      </Row>
    </>
  ))

  .add('distance and offset', () => (
    <>
      <h6>Distance</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Popover
            target={<Tag color="success" text="far away" />}
            distance={15}
          >
            far away
          </Popover>

          <Popover target={<Tag color="success" text="default offset" />}>
            default distance
          </Popover>

          <Popover
            target={<Tag color="success" text="very close" />}
            distance={3}
          >
            very close
          </Popover>
        </Col>
      </Row>

      <h6>Offset</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Popover
            target={<Tag color="success" text="positive offset" />}
            offset={100}
          >
            positive offset
          </Popover>

          <Popover target={<Tag color="success" text="default offset" />}>
            default offset
          </Popover>

          <Popover
            target={<Tag color="success" text="negative offset" />}
            offset={-100}
          >
            negative offset
          </Popover>
        </Col>
      </Row>
    </>
  ))

  .add('custom wrapper', () => (
    <>
      <h6>Custom wrapper</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Popover target="My target is in a <span>">
            By default, my target is wrapped in a span
          </Popover>

          <Popover target="My target is in a <div>" tag="div">
            You can change that with the tag property
          </Popover>
        </Col>
      </Row>
    </>
  ));

function TinyCrud() {
  const [open, setOpen] = useState(false);

  const persons = ['aap', 'noot', 'mies'];

  return (
    <ListGroup style={{ width: 300 }}>
      {persons.map((person) => (
        <ListGroupItem key={person}>
          <div className="d-flex justify-content-between">
            {person}

            <div>
              <Button
                icon="edit"
                className="mt-1"
                onClick={() => setOpen(true)}
              />

              <ConfirmButton
                dialogText="You sure son?"
                icon="delete"
                onConfirm={action('delete clicked')}
              />
            </div>
          </div>
        </ListGroupItem>
      ))}
      <ListGroupItem className="d-flex justify-content-center">
        <Button icon="add" className="mt-1" onClick={() => setOpen(true)}>
          Add person
        </Button>
      </ListGroupItem>

      {open ? (
        <OpenCloseModal onClose={() => setOpen(false)}>
          Form here?
        </OpenCloseModal>
      ) : null}
    </ListGroup>
  );
}

function NiceCard() {
  return (
    <Card>
      <CardHeader>This is a nice header</CardHeader>
      <CardBody>
        Lorem ipsum, dolor sit amet consectetur adipisicing elit. Provident
        eveniet, earum corrupti dicta quidem excepturi cupiditate consequuntur
        soluta obcaecati alias nobis eaque magnam sed et fugiat facere cumque,
        quaerat laborum!
      </CardBody>
      <CardFooter>Containing a footer</CardFooter>
    </Card>
  );
}
