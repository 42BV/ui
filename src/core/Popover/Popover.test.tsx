import React from 'react';
import { render, screen } from '@testing-library/react';

import { Popover } from './Popover';

describe('Component: Popover', () => {
  describe('ui', () => {
    test('default', () => {
      const { container } = render(
        <Popover
          target={
            <div>The popover should be wrapped around this div, in a span</div>
          }
        >
          Popover content
        </Popover>
      );
      expect(container).toMatchSnapshot();
    });

    test('open', () => {
      const { container } = render(
        <Popover
          isOpen={true}
          target={
            <div>The popover should be wrapped around this div, in a span</div>
          }
        >
          Popover content
        </Popover>
      );
      expect(container).toMatchSnapshot();
    });

    test('with custom tag', () => {
      render(
        <Popover tag="div" target="target">
          Popover content
        </Popover>
      );
      expect(screen.getByText('target').tagName).toBe('DIV');
    });

    test('with class', () => {
      render(
        <Popover className="extra-classnames" target="target">
          Popover content
        </Popover>
      );
      expect(
        screen.getByText('target').classList.contains('extra-classnames')
      ).toBe(true);
    });

    test('with optional properties', () => {
      const { container } = render(
        <Popover
          style={{ marginTop: 5, padding: 10 }}
          isOpen={true}
          distance={10}
          placement="bottom"
          onClickOutside={vi.fn()}
          target={
            <div>The popover should be wrapped around this div, in a div</div>
          }
        >
          Popover content
        </Popover>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
