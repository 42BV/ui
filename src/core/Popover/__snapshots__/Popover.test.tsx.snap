// Vitest Snapshot v1

exports[`Component: Popover > ui > default 1`] = `
<div>
  <span
    aria-expanded="false"
    role="button"
    tabindex="0"
  >
    <div>
      The popover should be wrapped around this div, in a span
    </div>
  </span>
</div>
`;

exports[`Component: Popover > ui > open 1`] = `
<div>
  <span
    aria-expanded="true"
    role="button"
    tabindex="0"
  >
    <div>
      The popover should be wrapped around this div, in a span
    </div>
  </span>
  <div
    data-tippy-root=""
    id="tippy-2"
    style="z-index: 1049; visibility: visible; transition: none; position: absolute; left: 0px; top: auto; margin: 0px; right: auto; bottom: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="tippy-box border-0 tippy-popover"
      data-animation="fade"
      data-escaped=""
      data-placement="top"
      data-reference-hidden=""
      data-state="hidden"
      role="tooltip"
      style="max-width: 350px; transition-duration: 0ms;"
      tabindex="-1"
    >
      <div
        class="tippy-content"
        data-state="hidden"
        style="transition-duration: 0ms;"
      >
        <div>
          Popover content
        </div>
      </div>
      <div
        class="tippy-arrow"
        style="position: absolute; left: 0px; transform: translate(0px, 0px);"
      />
    </div>
  </div>
</div>
`;

exports[`Component: Popover > ui > with optional properties 1`] = `
<div>
  <span
    aria-expanded="true"
    role="button"
    style="margin-top: 5px; padding: 10px;"
    tabindex="0"
  >
    <div>
      The popover should be wrapped around this div, in a div
    </div>
  </span>
  <div
    data-tippy-root=""
    id="tippy-5"
    style="z-index: 1049; visibility: visible; transition: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: auto; bottom: auto; transform: translate(0px, 0px);"
  >
    <div
      class="tippy-box border-0 tippy-popover"
      data-animation="fade"
      data-escaped=""
      data-placement="bottom"
      data-reference-hidden=""
      data-state="hidden"
      role="tooltip"
      style="max-width: 350px; transition-duration: 0ms;"
      tabindex="-1"
    >
      <div
        class="tippy-content"
        data-state="hidden"
        style="transition-duration: 0ms;"
      >
        <div>
          Popover content
        </div>
      </div>
      <div
        class="tippy-arrow"
        style="position: absolute; left: 0px; transform: translate(0px, 0px);"
      />
    </div>
  </div>
</div>
`;
